CodeSchool API Endpoints for CRUD & Rake Task for seeding from code_schools.yml

PrimeTimeTran opened the pull request:
# Description of changes
#76 
Created CRUD API endpoints for the `CodeSchool` model. Also added `Location` model as well because some of the CodeSchools had multiple locations.

Also created rake task for seeding data.
`rake schools:populate`            
Reads from the ./config/code_schools.yml and creates records in the database.

hpjaj commented on config/database.yml:
```diff
@@ -21,7 +21,8 @@ default: &default
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
   user: postgres
-  host: operationcode-psql
+  # host: operationcode-psql
```
@rickr - thoughts on this change?

PrimeTimeTran replied:
@hpjaj I just made this change to get it working on my local machine. I'm not sure how I could have gotten it to run the server otherwise.

hpjaj commented on lib/tasks/schools_from_yml.rake:
```diff
@@ -0,0 +1,30 @@
+        online_only: school["online_only"],
+        notes: school["notes"]
+        )
+      object.save
```
Perhaps consider using the `create!` method for creating the `CodeSchool` and `Location` records.  This will prevent any silent failures, and force the raising of an error, should there be an issue.  Then you can `rescue` and `log` those errors.  

Then you'll be able to identify any records that were not created, so you can go back and troubleshoot, manually add them, etc.

Here is [an example](https://github.com/OperationCode/operationcode_backend/blob/master/lib/tasks/resource.rake) of this.

hpjaj commented on test/controllers/api/v1/code_schools_controller_test.rb:
```diff
@@ -0,0 +1,20 @@
+    assert_response :missing
+  end
+end
+
```
Perhaps add in some test coverage for the `create`, `update`, and `destroy` actions?  Here are [some examples](https://github.com/OperationCode/operationcode_backend/blob/master/test/controllers/api/v1/team_members_controller_test.rb).

hpjaj commented on test/factories/code_schools.rb:
```diff
@@ -0,0 +1,12 @@
+FactoryGirl.define do
+  factory :code_school do
+    name "CoderSchool"
```
Perhaps leverage the [Faker gem](https://github.com/OperationCode/operationcode_backend/blob/master/test/factories/users.rb) to create unique `name`s, `url`s, `addresses`(in the other `location` factory). 

Here are [some examples](https://github.com/OperationCode/operationcode_backend/blob/master/test/factories/users.rb#L4-L12) of the gem being used.

hpjaj replied:
Can you update these with `Faker`.

hpjaj commented on test/models/code_school_test.rb:
```diff
@@ -0,0 +1,15 @@
+require 'test_helper'
+
+class CodeSchoolTest < ActiveSupport::TestCase
+  def setup 
```
Since the `code_school.rb` model file does not have any methods in it yet, the only thing you can really test for is validations.  So until we have some methods in there, you can probably just delete this file.

hpjaj commented on test/models/location_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class LocationTest < ActiveSupport::TestCase
```
Same thing here.

hpjaj commented on config/application.rb:
```diff
@@ -33,5 +33,9 @@ class Application < Rails::Application
 
     config.secret_path = Rails.root.join('run/secrets')
+    
+    Raven.configure do |config|
```
Can you elaborate on why this is being added?

PrimeTimeTran replied:
I added this because I got an error `sentry-raven DSN not set`. This was the solution I found on the internet to get the server to run locally. I plan on dropping it when I finish the PR.

hpjaj commented on app/controllers/api/v1/code_schools_controller.rb:
```diff
@@ -2,7 +2,50 @@ module Api
+        end
+      end
+      
+      def destroy
```
Please be sure to update our API documentation with these new endpoints (`index, create, show, update, destroy`).  

You'll do so by updating the  [apiary.apib](https://github.com/OperationCode/operationcode_backend/blob/master/apiary.apib) file, following the pattern in the `TeamMember` section.

Here is what that file yields:  http://docs.operationcodeapi.apiary.io/#

PrimeTimeTran commented on app/models/code_school.rb:
```diff
@@ -0,0 +1,4 @@
+class CodeSchool < ApplicationRecord
+  validates :name, :url, :logo, :full_time, :hardware_included, :has_online, :online_only, presence: true
```
@hpjaj I'm not sure what OperationCode would like to validate so I figured I just add them all and delete before merging the PR. Concerning this question, yes I did run the rake task locally and the only field that was not on every school was `notes`

PrimeTimeTran commented:
@hpjaj could you take a look at the API Documentation I've added? Besides this I'll remove the configuration for `Raven` I added in `./config/application.rb` and the change the `host:` back to `host: operationcode-psql` in `./config/database.yml` and I think it'll be close to done. 

Is there anything else you can see that I've missed?

hpjaj commented on lib/tasks/schools_from_yml.rake:
```diff
@@ -0,0 +1,28 @@
+  task populate: :environment do 
+    schools =  YAML::load_file(File.join("./config", "code_schools.yml"), "r")    
+    schools.each do |school|
+      object = CodeSchool.create!(
```
Can you add in a `rescue` block so that we catch any errors, and the rest of the schools that are error free can be created.

hpjaj commented on test/controllers/api/v1/code_schools_controller_test.rb:
```diff
@@ -0,0 +1,51 @@
+    get api_v1_code_school_path(@school), as: :json
+    assert_response :ok
+  end
+  
```
Can you add in a test for `:create`.

PrimeTimeTran replied:
@hpjaj coulda swore I wrote one, musta overwritten it somewhere.

hpjaj commented on test/factories/locations.rb:
```diff
@@ -0,0 +1,11 @@
+FactoryGirl.define do
+  factory :location do
+    va_accepted false
+    address1 "MyString"
```
Here too, removing any `MyString`.

PrimeTimeTran replied:
@hpjaj if I use Faker I'd have to add some more code to the tests, do you want me to user faker on all of the attributes?

hpjaj replied:
Adding more code to the tests is fine.  The more representative our tests are of the real world, the better.

Yes, please use faker in situations where uniqueness is valuable, and more representative of the real world.  

You can still choose specific attributes in the tests if needs be, i.e.

```ruby
let(:location) { create :location, address1: "1515 Main St.", .. }
```

PrimeTimeTran replied:
Ok let me do all of this and ping you once I've completed. Thanks for your help~!

hpjaj commented on app/controllers/api/v1/code_schools_controller.rb:
```diff
@@ -2,7 +2,50 @@ module Api
+        if school.save 
+          render json: { code_school: school.id }
+        else 
+          render json: { :errors => school.errors.full_messages }
```
Can you change these to the newer `errors: school.errors.full_messages` syntax.

PrimeTimeTran replied:
@hpjaj yeah, sorry about that

hpjaj commented on apiary.apib:
```diff
@@ -577,3 +577,191 @@ API endpoints that Operation Code's Rails backend makes available to its React f
             errors: "Some error message"
         }
+
+## CodeSchool | Collection [/api/v1/code_schools]
```
Two things:

1.  Please create a free account at apiary.io, and use their editor to confirm that all of these changes do not create any semantic issues 
2. These endpoints are alphabetized, so these additions will need to go at the top of the file

PrimeTimeTran commented:
@hpjaj I've worked on the requests you had.

hpjaj commented on apiary.apib:
```diff
@@ -5,6 +5,195 @@ HOST: https://operationcode.org/
+        ]
+        
+        
+### CodeSchool | Create [/api/v1/code_schools{?name,url,logo,full_time,hardware_included,has_online,online_only,notes}]
```
I believe we'll want this to be an `h2` header instead of an `h3`, with an `h3` beneath it, because it is its own endpoint.

hpjaj commented on apiary.apib:
```diff
@@ -5,6 +5,195 @@ HOST: https://operationcode.org/
+              errors: "Some error message"
+          }
+          
+## CodeSchool | Show [/api/v1/code_schools/{code_school_id}] [GET]
```
We'll want an `h3` header below this one, so that it stands alone as its own endpoint.  

The way apiary enforces formatting, we essentially need to have one `h2` and one `h3` for each new endpoint.

PrimeTimeTran replied:
@hpjaj I get complaints from Apiary when I try to format it the way you suggested. 
I don't have another example of a resource with a `show` & `delete` endpoint to reference on how to get by this. 

Do you know of a workaround? Line `92` & `118`

![screen shot 2017-09-27 at 12 48 03 am](https://user-images.githubusercontent.com/16382165/30875475-dc6df20a-a31d-11e7-9f50-eb14c8ef8a96.jpg)
 so I don't get the semantic issue?

hpjaj replied:
For the case where both API endpoints are identical, and have the same parameters, i.e. `[/api/v1/code_schools/{code_school_id}]`, I would: 

- have an `h2` that addresses both (i.e. `## CodeSchool | Show or Delete`)
- have an `h3` for each endpoint, beneath that `h2`

PrimeTimeTran replied:
@hpjaj thanks for the support on completing this task. Mind taking a look again? I've updated with all of your suggestions.

hpjaj commented on lib/tasks/schools_from_yml.rake:
```diff
@@ -0,0 +1,32 @@
+          )
+        end
+      rescue ActiveRecord::RecordInvalid => e
+        puts "Failed to create CodeSchool: #{e}"
```
Can you add a `Rails.logger.error ...` so that we capture the error in the logs, too.

PrimeTimeTran replied:
Done.

hpjaj commented on test/factories/locations.rb:
```diff
@@ -0,0 +1,11 @@
+FactoryGirl.define do
+  factory :location do
+    va_accepted Faker::Boolean.boolean
```
In the other factories, the call to `Faker` is wrapped in `{ }`:  https://github.com/OperationCode/operationcode_backend/blob/master/test/factories/users.rb#L4

Can you confirm if these are needed or not?

PrimeTimeTran replied:
@hpjaj I'm reasonably confident we don't need the `{}`'s, all the tests pass when using the `create(:code_school)` method, where we see the Factory is defined using the same syntax.

PrimeTimeTran commented:
@hpjaj I've committed and pushed the changes you requests =).

hpjaj commented:
@PrimeTimeTran - There ended up being a couple issues with this PR.  I wanted to chat with you about them.

Are you a member of the Operation Code Slack channel?  If yes, what is your @ name?

If no, you can join by going to https://operationcode.org/profile and clicking **Enter our Slack channel**.  

Once you are in Slack, my handle is **@ harry**.

PrimeTimeTran commented:
@hpjaj I see, sorry I missed them =(.

No, I'm not. ok I'll join =)
